aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -226,12 +224,116 @@ Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim) {
+  }
+  dim = at::maybe_wrap_dim(dim, input.dim());
+  const int64_t dim_size = input.size(dim);
+  if (dim_size == 1) {
+    return grad;
+  }
 
+  const auto w = output * grad;
+  const auto is_zero = input == 0;
+  if (!(is_zero.any().item<uint8_t>())) {
+    return reversed_cumsum(w, dim).div(input);
+  }
+
+  // If we are not computing a second order gradient, we can use an
+  // O(n) implementation. The derivative of this implementation is _not_
+  // the second derivative of cumprod. As such, we fallback to a less efficient
+  // O(n^2) implementation when at::GradMode::is_enabled().
+  Tensor grad_input = at::zeros(input.sizes(), grad.options());
+  if (!at::GradMode::is_enabled()) {
+    // n.b. This could probably be implemented much faster with a kernel
+
+    // From here on we need to use some mask gymnastics to
+    // account for the tensorial dimensions
+    // We do a cumsum of the zeros along the dimension.
+    // For a vector is_zero = [False, True, Fals, True, False]
+    // we would have cumsum = [0, 1, 1, 2, 2]
+    // As such we have (in python code for simplicity)
+    // The mask for the range [0, z1):
+    // cumsum == 0
+    // The indices of the first zero z1 and zeros when
+    // there is no first zero:
+    // indices = (cumsum == 1).max(dim, keepdim=True).indices
+    // The mask for the first zero:
+    // zeros_like(indices).scatter_(dim, indices, 1.) & cumsum == 1
+    // Note that the logic_and with cumsum == 1 accounts
+    // for the case when there is no first zero
+    const auto cumsum = is_zero.cumsum(dim);
+
+    // case k < z1
+    // select everything before the first zero [0, z1)
+    auto mask = cumsum == 0;
+    // equiv to grad_input[mask] = deriv[grad]
+    grad_input.masked_scatter_(mask,
+        reversed_cumsum(w.masked_fill(~mask, 0.), dim).div_(input).masked_select(mask));
+    // select everything from the first zero to the second zero [z1, z2)
+    mask = cumsum == 1;
+
+    // case k = z1
+    // We start by select the first zero [z1]
+    // We locate the indices of the first zero using the max function
+    // We then go from the indices to a mask using scatter_
+    // When there is no zero in the slice, max will return the index 0.
+    // To account for this, we need to apply the mask which is true in the range [z1, z2)
+    const auto first_zero_index = std::get<1>(mask.max(dim, /*keepdim*/ true));
+    const auto first_zero_mask = at::zeros_like(mask)
+                                  .scatter_(dim, first_zero_index, /*src*/ true)
+                                  .logical_and_(mask);
+    // select everything between the first zero and the second zero (z1, z2)
+    mask &= ~first_zero_mask;
+    // here we compute
+    // dy_j / dx_z1 = sum(cumprod(input[z1+1:z2] * grad[z1+1:z2])) * prod(output[z1-1])
+    // relu_() necessary as gather does not support negative indices
+    // finally, we do grad_input[z1] = dy_j / dx_z1
+    grad_input.masked_scatter_(first_zero_mask,
```

Comment:
we should really have a separate kernel! we are calling into 12 kernels just here ...

Reply:
I agree that a separate kernel would make things more efficient. That being said, I do not think that this function is so commonly used as to be worth the effort.

I left that as a comment, in case anyone wants to take on it in the future.

Reply:
yeah I agree. could you benchmark the performance for this case and ensure that we are not causing any regression though?

Reply:
It is benchmarked under the `zeros` paragraph within the benchmarks. In fact, it is in these cases when we get the 7-30x speed-up on GPU.

Reply:
perfect thanks!!